### DISCUSSION TITLE
Does State.Vars store values in the backend or frontend ?

### DISCUSSION THREAD

**Discussion by yellowbean** [reflex-dev/reflex#3002] [Q&A] [answered]
Great framework ! Thank you guys !

I would like to store a field of State with large values ( probably a dict of map ( dict[str,str] )  with 100 MB.

Is the var is being store in backend ( thus being hold in memory ?) or the var is being stored at frontend ,like the browser ?

**Comment by a0s** [comment]
And will it be placed inside redis? The default limit on value in Redis is 500MB btw..

**Comment by picklelo** [comment]
See this [blog post](https://reflex.dev/blog/2024-03-21-reflex-architecture/#the-reflex-architecture) for how Reflex works under the hood.

The state vars will be stored on the backend and will take up memory. If you're using Redis then it will be stored there. What's your use case where you need such a large state var? Is this var independent for each user of your app, or would it be possible to use a database to store this information?

**Reply by yellowbean** [reply]
Thanks @picklelo 

- large vars stores large dataframes/jsons for user to view them.( probably render with DataTable ) 
- this var independent for each user 

User will use these data on the fly ,so it doesn't have to be persistent in database. ( These data could be saved in database, but I'm not sure why as once user log out ,then data is no longer required, not sure whether it is worth to have database IO for this.

**Reply by picklelo** [reply]
Got it, yes unfortunately it will create a copy of the data frame for each new user. I think there's no real way around this without using an external store, as the different dataframes need to be stored on the backend together, and it would be impractical to send the data back and forward from the frontend for processing. As long as the number of users to the app are limited it should work, but it won't scale to many many users.

**Reply by yellowbean** [reply]
Thanks @picklelo , that's very helpful introduction to understand `reflex`. so `state manager` will mapping a single user to a `state` ? 
i.e. If a user create a big var in the state. then the var will be active in the backend server forever till the fastapi server is killed ?  or the var will be wipe out in the backend once the browser tab is closed ?

**Reply by picklelo** [reply, marked as answer]
Yes the state manager is pretty much a dictionary from a user token to their state. If you're using the in-memory state manager then yes the state will be there until the server is killed, but if you use Redis by setting a `redis_url`in the `rxconfig.py`, as we do in our hosting service, then there is an expiration that will wipe it out.

We can add an issue for implementing an expiration for the memory state manager. We couldn't do it on the socket disconnect as then if someone has intermittent connection issues. I made this issue as many of our Docker examples don't use Redis and can lead people to this issue: https://github.com/reflex-dev/reflex/issues/3021
